packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+  protected unpatch(): void {
+    shimmer.unwrap(this._moduleExports, 'request');
+    if (semver.satisfies(this.version, '>=8.0.0')) {
+      shimmer.unwrap(this._moduleExports, 'get');
+    }
+    if (
+      this._moduleExports &&
+      this._moduleExports.Server &&
+      this._moduleExports.Server.prototype
+    ) {
+      shimmer.unwrap(this._moduleExports.Server.prototype, 'emit');
+    }
+  }
+
+  /**
+   * Creates spans for incoming requests, restoring spans' context if applied.
+   */
+  protected _getPatchIncomingRequestFunction() {
+    return (original: (event: string, ...args: unknown[]) => boolean) => {
+      return this._incomingRequestFunction(original);
+    };
+  }
+
+  /**
+   * Creates spans for outgoing requests, sending spans' context for distributed
+   * tracing.
+   */
+  protected _getPatchOutgoingRequestFunction() {
+    return (original: Func<ClientRequest>): Func<ClientRequest> => {
+      return this._outgoingRequestFunction(original);
+    };
+  }
+
+  protected _getPatchOutgoingGetFunction() {
+    return (original: Func<ClientRequest>): Func<ClientRequest> => {
+      // Re-implement http.get. This needs to be done (instead of using
+      // getPatchOutgoingRequestFunction to patch it) because we need to
+      // set the trace context header before the returned ClientRequest is
+      // ended. The Node.js docs state that the only differences between
+      // request and get are that (1) get defaults to the HTTP GET method and
+      // (2) the returned request object is ended immediately. The former is
+      // already true (at least in supported Node versions up to v10), so we
+      // simply follow the latter. Ref:
+      // https://nodejs.org/dist/latest/docs/api/http.html#http_http_get_options_callback
+      // https://github.com/googleapis/cloud-trace-nodejs/blob/master/src/plugins/plugin-http.ts#L198
+      return function outgoingGetRequest<
+        T extends RequestOptions | string | URL
+      >(options: T, ...args: HttpRequestArgs) {
+        const req = request(options, ...args);
+        req.end();
+        return req as ClientRequest;
+      };
+    };
+  }
+
+  /**
+   * Injects span's context to header for distributed tracing and finishes the
+   * span when the response is finished.
+   * @param request The original request object.
+   * @param options The arguments to the original function.
+   * @param span representing the current operation
+   */
+  private _getMakeRequestTraceFunction(
+    request: ClientRequest,
+    options: ParsedRequestOptions,
+    span: Span
+  ): Func<ClientRequest> {
+    return (): ClientRequest => {
+      this._logger.debug('makeRequestTrace by injecting context into header');
+
+      const propagation = this._tracer.getHttpTextFormat();
+      const opts = Utils.getIncomingOptions(options);
+
+      propagation.inject(span.context(), Format.HTTP, opts.headers);
+
+      request.on('response', (response: IncomingMessage) => {
```

Comment:
installing a 'response' listener has side effects. If there is no listener Node will automatically discard any incoming data but if there is one it's up to the listener to consume the data. Therefore depending on the number of listeners we should consume the data otherwise connections may hang.

Reply:
created https://github.com/open-telemetry/opentelemetry-js/issues/225